### ISSUE TITLE
URL not loading because missing username and password on HTTP Authentication

### ISSUE THREAD

**Issue by gatuno** [kodi-pvr/pvr.iptvsimple#715]
Hello

I'm not sure if the bug is on the pvr.iptvsimple addon or kodi interface, but I'm working on kodi Nexus 20.0.0, and the playlist I'm trying to load is a M3U protected by a basic HTTP Auth password, and it fails to load the playlist with a 401 error. I enabled the debug log as you recommend on other issues, and is available here: https://paste.kodi.tv/abepebefaf.kodi

Also, I kinda debug the source code from the iptvsimple to understand the issue, and looks like it starts with PlaylistLoader::LoadPlayList, and as I don't use cache, it calls FileUtils::GetCachedFileContents, next it calls FileUtils::GetFileContents where the parameter url still has the format of "http://USER:PASS@url.com/folder/list.m3u" and it fails with a 401 http code. Looks like either kodi changed his api to pass a user/password.
From the debug file, I can see that:

`CCurlFile::Open - <http://domain.somewhere.net/folder/list.m3u> Failed with code 401:`

I'm comparing the debug from a kodi v19, and looks like on the CCurlFile::Open it passes the user and password correctly, and then it loads correctly.

It could be pretty useful if the http authentication could be used. Or maybe, a config where the user and password can be written.

Thanks for reading.

**Comment by phunkyfish** [comment]
You need to embed the username/password in the URLs in you M3U file.

Or are you saying that the URLs contain the username/password and that they are being stripped somehow?

**Comment by phunkyfish** [comment]
You mean for the loading of the playlist itself. Ok, that has always worked. And you are sure that in the addon settings the username and password are part of the playlist URL?

**Comment by gatuno** [comment]
Yes, the URL for loading the playlist itself. I know what "ilogical" what this bug sounds, but is easy to reproduce. Setup a m3u file behind a basic http authentication, and try to load it from the plugin using the syntax "http://USER:PASS@domain.com/folder/file.m3u"

**Comment by phunkyfish** [comment]
The curl log that you see will strip out the username and password so it’s not printed in the log. Are we sure it’s not the user agent that is causing the problem?

**Comment by gatuno** [comment]
In Kodi v19 curl log hides the user and password. as I have control of the http server, I am going to check if both request are made correctly or just the first one without authentication.

Also, not a user-agent related problem because it's a simple static file on a personal web server, no "special" configs or problems. Just a plain basic http auth

**Comment by vpeter4** [comment]
Just made quick test and playlist "http://USER:PASS@url.com/folder/list.m3u" was loaded.
You should enable Kodi logging with Curl component and you should see 2 HTTP requests. First is without authentication and 401 code is returned. Second request is made with basic authentication which should succeed if username and password match.

**Comment by gatuno** [comment]
I finally had the time to do the testing.

I find out the issue. The url does work, but it does replies with a "301 Moved Permanently" response. In Kodi 20, it doesn't follow the Location header. But in kodi 19 it did followed it and after that it does the authentication stuff.

Also, its a redirect from the plain http to the secure https. If I put the url with the https directly it correctly authtenticates with no issues.

So to be clear, the issue happens if there is a "301" response before the 401 code.

**Comment by phunkyfish** [comment]
Hmmm, this feels like it might be kodi core issue but I’m not sure at all.

**Comment by vpeter4** [comment]
Just made a test and everything works well here.
```
LibreELEC (community): nightly-20221224-6488140 (Generic.x86_64)
LibreELEC (community): nightly-20230312-5bf7662 (Generic.x86_64)

Kodi (20.1 (20.1.0) Git:20230312-289ec664e3). Platform: Windows NT x86 64-bit
Kodi compiled 2023-03-12 by MSVC 193431937 for Windows NT x86 64-bit version 10.0 (0x0A00000B)
```

GET, 401, GET with Authorization, 301, 206 with data

**Comment by gatuno** [comment]
I do think is a Kodi core issue, but didn't know where to start.

As @vpeter4 said, he got a 401 first, next the 301. It works.

In my case, it goes like this:
GET, 301, GET again, 401, GET with Authorization, 20x with data

This used to work on Kodi 19.

**Comment by vpeter4** [comment]
Will simulate this case tomorrow.

**Comment by vpeter4** [comment]
I know I have to leave my test setup intact :)

Anyway, everything still works for me with
GET, 301, GET, 401, GET with Authorization, 206 with data

@gatuno: Did you check with Wireshark that you really have GET requests and not only HEAD?
I had HEAD requests until I cleared iptv addon cache to really get new playlist.
